### PR TITLE
Cluster MSI follow-up PR

### PR DESCRIPTION
**Pull request by kimorris27**
### Which issue this PR addresses:

https://issues.redhat.com/browse/ARO-4360

### What this PR does / why we need it:

The main purpose of this PR is to remove our dependency on https://github.com/golang/mock and completely switch over to using https://github.com/uber-go/mock instead for two reasons:

1. It needs to be done eventually anyway
2. It happens to fix an issue that I ran into generating mocks for some new Azure clients I added in my previous cluster MSI PR (the golang version doesn't seem to be up-to-date enough to handle generics properly in certain contexts)

I also added in a few small changes in response to previous PR feedback. If you follow the commit history, it's hopefully easy to pick out which of the numerous changed files you need to care about as a reviewer (as opposed to the ones where the only change is the mock repo referenced). Notably, I upgrade Go from 1.21.11 to 1.21.13 to satisfy Uber mockgen's dependency on Go >= 1.21.12.

### Test plan for issue:

CI and E2E

### Is there any documentation that needs to be updated for this PR?

I don't think so, but let me know if you know of any.

### How do you know this will function as expected in production? 

E2E during the next release :D

### CODE REVIEW

**Comment by kimorris27** [comment]
E2E failed because our arointsvc registry doesn't have the Go 1.21.13 image. Looks like I'll need to follow what was done here to fix that: https://github.com/Azure/ARO-RP/pull/3698#issuecomment-2233560573

File: pkg/util/azureclient/azuresdk/armcosmos/generate.go
Comment (tsatam) [inline pkg/util/azureclient/azuresdk/armcosmos/generate.go:8]: If we weren't generating mocks for these two clients, that means we weren't using mock clients in our tests for RP functionality using these clients, correct? (assuming there even are tests). 

If that's the case, should we even bother generating mocks that go unused?

**Review by tsatam — APPROVED** [review]
I've reviewed the changes and they look good, but I'd suggest perhaps splitting the "replace golang/mock with uber-go/mock" part out into a separate PR to get more focused review from the wider team, only if it is easy enough to do so.

**Comment by kimorris27** [comment]
Closing this to split into two more granular PRs: one that moves us to the Uber mock fork and one that has the follow-up stuff from the previous cluster MSI PR.

File: pkg/util/azureclient/azuresdk/armcosmos/generate.go
Comment (kimorris27) [reply pkg/util/azureclient/azuresdk/armcosmos/generate.go:8]: So there were mocks present, but if you remove them and `make generate`, they don't get replaced. I'm guessing that when they were added, the committer forgot to commit their changes to the `generate.go` files.

This came to my attention as I was moving us over to the Uber fork because it caused errors while I was trying to get builds and unit tests working, so I codified the generation properly.

**Pull request closed without merging**
